spread/openstack.go

```diff
@@ -0,0 +1,640 @@
+	}
+
+	if len(zones) == 0 {
+		return nil, &FatalError{errors.New("cannot find any availability zones")}
+	}
+	return &zones[0], nil
+}
+
+func (p *openstackProvider) findSecurityGroupNames(names []string) ([]nova.SecurityGroupName, error) {
+	var secGroupNames []nova.SecurityGroupName
+
+	secGroups, err := p.networkClient.ListSecurityGroupsV2()
+	if err != nil {
+		return nil, fmt.Errorf("cannot retrieve security groups: %s", errorMsg(err))
+	}
+
+	if len(secGroups) == 0 {
+		return nil, &FatalError{errors.New("cannot find any groups")}
+	}
+
+	for _, name := range names {
+		found := false
+		for _, sg := range secGroups {
+			if name == sg.Name {
+				found = true
+				break
+			}
+		}
+		if !found {
+			return nil, &FatalError{fmt.Errorf("cannot find valid group with name %s", name)}
+		}
+		secGroupNames = append(secGroupNames, nova.SecurityGroupName{Name: name})
+	}
+	return secGroupNames, nil
+}
+
+func (p *openstackProvider) waitServerCompleteBuilding(s *openstackServer, timeoutSeconds int) error {
+	// Wait until the server is actually running
+	start := time.Now()
+	for {
+		if time.Since(start) > time.Duration(timeoutSeconds)*time.Second {
+			return &FatalError{fmt.Errorf("cannot check status: timeout reached")}
+		}
+		server, err := p.computeClient.GetServer(s.d.Id)
+		// When the server info cannot be retrieved, wait 2 seconds to retry
+		if err != nil {
+			time.Sleep(2 * time.Second)
+			continue
+		}
+		if server.Status != nova.StatusBuild {
+			if server.Status != nova.StatusActive {
+				return &FatalError{fmt.Errorf("cannot use server: status is not active but %s", server.Status)}
+			}
+			break
+		}
+		time.Sleep(5 * time.Second)
+	}
+	debugf("Server %s is running", s.d.Name)
+	return nil
+}
+
+func (p *openstackProvider) waitServerCompleteSetup(s *openstackServer, timeoutSeconds int) error {
+	server, err := p.computeClient.GetServer(s.d.Id)
+	if err != nil {
+		return fmt.Errorf("cannot retrieving server information: %s", errorMsg(err))
+	}
+	// The adreesses for a network is map of networks and list of ip adresses
+	// We are configuring just 1 network address for the network
+	// To get the address is used the first network
+	if len(server.Addresses) > 0 && len(s.d.Networks) > 0 {
+		s.address = server.Addresses[s.d.Networks[0]][0].Address
+	} else {
+		return fmt.Errorf("cannot retrieve server address")
+	}
+
+	config := &ssh.ClientConfig{
+		User:            "root",
+		Auth:            []ssh.AuthMethod{ssh.Password(p.options.Password)},
+		Timeout:         10 * time.Second,
+		HostKeyCallback: ssh.InsecureIgnoreHostKey(),
+	}
+	addr := s.address
+	if !strings.Contains(addr, ":") {
+		addr += ":22"
+	}
+
+	// Iterate until the ssh connection to the host can be stablished
+	// In openstack the client cannot access to the serial console of the instance
+	start := time.Now()
+	for {
+		if time.Since(start) > time.Duration(timeoutSeconds)*time.Second {
```

Comment:
same comment as above regarding following timeout pattern in `googleProvider.waitOperation`.